File: call-notes-to-pedigree/node/src/httpGateway.ts
/**
 * HTTP gateway abstraction and a `fetch`-backed implementation.
 */

export interface HttpResponse {
  readonly status: number;
  text(): Promise<string>;
  json(): Promise<unknown>;
}

export interface HttpGatewayOptions {
  readonly method: 'POST' | 'PATCH' | 'DELETE';
  readonly url: string;
  readonly headers: Record<string, string>;
  readonly body: unknown;
}

export interface HttpGateway {
  send(options: HttpGatewayOptions): Promise<HttpResponse>;
}

export class FetchHttpGateway implements HttpGateway {
  constructor(private readonly timeoutMs = 30_000) {}

  async send(options: HttpGatewayOptions): Promise<HttpResponse> {
    const controller = new AbortController();
    const timer = setTimeout(() => {
      controller.abort();
    }, this.timeoutMs);
    try {
      const response = await fetch(options.url, {
        method: options.method,
        headers: options.headers,
        body: JSON.stringify(options.body),
        signal: controller.signal,
      });
      const text = await response.text();
      return {
        status: response.status,
        text: () => Promise.resolve(text),
        json: () => Promise.resolve(text === '' ? null : (JSON.parse(text) as unknown)),
      };
    } finally {
      clearTimeout(timer);
    }
  }
}

File: family-history-intake-form/python/src/family_intake/http_gateway.py
"""HTTP gateway abstraction and an httpx-backed implementation."""

from __future__ import annotations

from typing import Any, Literal, Protocol

import httpx

HttpMethod = Literal["POST", "PATCH"]


class HttpResponse(Protocol):
    status_code: int

    def json(self) -> Any: ...


class HttpGateway(Protocol):
    def send(
        self,
        method: HttpMethod,
        url: str,
        *,
        headers: dict[str, str],
        body: dict[str, Any],
    ) -> HttpResponse: ...


class HttpxGateway:
    """Concrete :class:`HttpGateway` backed by ``httpx.Client``."""

    def __init__(self, timeout_seconds: float = 10.0) -> None:
        self._client = httpx.Client(timeout=timeout_seconds)

    def send(
        self,
        method: HttpMethod,
        url: str,
        *,
        headers: dict[str, str],
        body: dict[str, Any],
    ) -> httpx.Response:
        return self._client.request(method, url, headers=headers, json=body)

    def close(self) -> None:
        self._client.close()
